File: tici.py
#!/usr/bin/env python3
"""Query TiCI metadata tables."""

import argparse
import shlex
import subprocess
import sys
from dataclasses import dataclass
from typing import List, Optional, Sequence


DEFAULT_HOST = "10.2.12.81"
DEFAULT_PORT = 9529
DEFAULT_USER = "root"
DEFAULT_DATABASE = "tici"
DEFAULT_META_TABLE = "tici_shard_meta"


def quote_identifier(identifier: str) -> str:
    if not identifier:
        raise ValueError("identifier must not be empty")
    return f"`{identifier.replace('`', '``')}`"


@dataclass
class MySQLClient:
    host: str = DEFAULT_HOST
    port: int = DEFAULT_PORT
    user: str = DEFAULT_USER
    database: str = DEFAULT_DATABASE
    password: str = ""
    mysql_bin: str = "mysql"
    comments: bool = True

    def command(self) -> List[str]:
        cmd = [
            self.mysql_bin,
            "--host",
            self.host,
            "--port",
            str(self.port),
            "-u",
            self.user,
        ]
        if self.comments:
            cmd.append("--comments")
        if self.password:
            cmd.append(f"-p{self.password}")
        cmd.extend(["-D", self.database])
        return cmd

    def execute(self, sql: str, *, echo: bool = False) -> None:
        cmd = self.command()
        if echo:
            print(f"$ {shlex.join(cmd)}")
            print(sql)
        subprocess.run(cmd, input=sql, text=True, check=True)


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(description="Query TiCI shard metadata")
    parser.add_argument("--host", default=DEFAULT_HOST, help=f"MySQL host, default: {DEFAULT_HOST}")
    parser.add_argument("--port", type=int, default=DEFAULT_PORT, help=f"MySQL port, default: {DEFAULT_PORT}")
    parser.add_argument("--user", default=DEFAULT_USER, help=f"MySQL user, default: {DEFAULT_USER}")
    parser.add_argument("--password", default="", help="MySQL password, optional")
    parser.add_argument("--database", default=DEFAULT_DATABASE, help=f"Database name for mysql session, default: {DEFAULT_DATABASE}")
    parser.add_argument("--mysql-bin", default="mysql", help="Path to mysql client binary")
    parser.add_argument(
        "--meta-table",
        default=DEFAULT_META_TABLE,
        help=f"Metadata table name, default: {DEFAULT_META_TABLE}",
    )
    parser.add_argument("--dry-run", action="store_true", help="Print SQL only, do not execute")
    return parser


def build_client(args: argparse.Namespace) -> MySQLClient:
    return MySQLClient(
        host=args.host,
        port=args.port,
        user=args.user,
        password=args.password,
        database=args.database,
        mysql_bin=args.mysql_bin,
    )


def build_query_sql(meta_table: str) -> str:
    target = f"{quote_identifier('tici')}.{quote_identifier(meta_table)}"
    return f"select table_id,index_id,shard_id,progress from {target};"


def main(argv: Optional[Sequence[str]] = None) -> int:
    parser = build_parser()
    args = parser.parse_args(list(argv) if argv is not None else None)

    try:
        sql = build_query_sql(args.meta_table)
        if args.dry_run:
            print(sql)
            return 0
        build_client(args).execute(sql)
        return 0
    except subprocess.CalledProcessError as exc:
        print(f"command failed with exit code {exc.returncode}", file=sys.stderr)
        return exc.returncode
    except ValueError as exc:
        print(str(exc), file=sys.stderr)
        return 2


if __name__ == "__main__":
    raise SystemExit(main())

File: src/bin/tici.rs
use clap::Parser;
use tici_test_rust::common::{MysqlConfig, print_stderr_log, quote_identifier};

const DEFAULT_HOST: &str = "10.2.12.79";
const DEFAULT_PORT: u16 = 9528;
const DEFAULT_USER: &str = "root";
const DEFAULT_DATABASE: &str = "tici";
const DEFAULT_META_TABLE: &str = "tici_shard_meta";

#[derive(Parser, Debug)]
#[command(about = "Query TiCI shard metadata")]
struct Cli {
    #[arg(long, default_value = DEFAULT_HOST)]
    host: String,
    #[arg(long, default_value_t = DEFAULT_PORT)]
    port: u16,
    #[arg(long, default_value = DEFAULT_USER)]
    user: String,
    #[arg(long, default_value = "")]
    password: String,
    #[arg(long, default_value = DEFAULT_DATABASE)]
    database: String,
    #[arg(long, default_value = "mysql")]
    mysql_bin: String,
    #[arg(long, default_value = DEFAULT_META_TABLE)]
    meta_table: String,
    #[arg(long, default_value_t = false)]
    dry_run: bool,
}

fn build_query_sql(meta_table: &str) -> Result<String, String> {
    let target = format!(
        "{}.{}",
        quote_identifier("tici")?,
        quote_identifier(meta_table)?
    );
    Ok(format!(
        "select table_id,index_id,shard_id,progress from {target};"
    ))
}

fn main() {
    let cli = Cli::parse();
    let sql = match build_query_sql(&cli.meta_table) {
        Ok(value) => value,
        Err(err) => {
            print_stderr_log(&err);
            std::process::exit(2);
        }
    };
    if cli.dry_run {
        println!("{sql}");
        return;
    }
    let client = MysqlConfig {
        host: cli.host,
        port: cli.port,
        user: cli.user,
        password: cli.password,
        database: cli.database,
        mysql_bin: cli.mysql_bin,
        comments: true,
    };
    if let Err(err) = client.execute_cli_stdin(&sql, false) {
        print_stderr_log(&err);
        std::process::exit(1);
    }
}
